document/document.go

```diff
@@ -335,6 +335,30 @@ func (d *Document) insertTable(relativeTo Paragraph, before bool) Table {
 	return d.AddTable()
 }
 
+func (d *Document) tables(bc *wml.EG_ContentBlockContent) []Table {
+	ret := []Table{}
+	for _, t := range bc.Tbl {
+		ret = append(ret, Table{d, t})
+		for _, crc := range t.EG_ContentRowContent {
+			for _, tr := range crc.Tr {
+				for _, ccc := range tr.EG_ContentCellContent {
+					for _, tc := range ccc.Tc {
+						for _, ble := range tc.EG_BlockLevelElts {
+							for _, cbc := range ble.EG_ContentBlockContent {
+								for _, tbl := range d.tables(cbc) {
```

Comment:
Is there a possibility that we may run into an infinite recursion, maybe with the mailformed doc? Maybe we should detect this kind of situation and handle it someway?

Reply:
This comment relates to this: https://github.com/baliance/gooxml/pull/255#discussion_r270619267

I have a few thoughts about this.

1. There is at least once other place in the library where you can run into the same issue, the `document.Save` and `document.SaveToFile` functions.
2. It's actually pretty difficult to create the infinite recursion since you have to explicitly construct `EG_BlockLevelElts` and `EG_ContentBlockContent` for two separate objects, passing parent/child into one and the other.
3. The stdlib bas this same issue with the encoding package. The solution there was to do nothing: https://github.com/golang/go/issues/10769

My two prefered solutions would be to either develop a function that will check the entire document for pointer loops, or do nothing.

If a separate function is going to be developed, that is probably outside the scope of this PR.

Here is the example I created to create infinite recursion:

```go
package main

import (
	"baliance.com/gooxml/document"
	"baliance.com/gooxml/schema/soo/wml"
)

func main() {
	doc := document.New()

	tbl1 := doc.AddTable()
	tbl2 := doc.AddTable()
	cell1 := tbl1.AddRow().AddCell()
	cell2 := tbl2.AddRow().AddCell()

	// tbl1 -> cell1 -nests-> tbl2 -> cell2 -nests-> tbl1

	// place tbl2 inside cell tbl1/cell1
	tc1 := cell1.X()
	elts := wml.NewEG_BlockLevelElts()
	tc1.EG_BlockLevelElts = append(tc1.EG_BlockLevelElts, elts)
	c := wml.NewEG_ContentBlockContent()
	elts.EG_ContentBlockContent = append(elts.EG_ContentBlockContent, c)
	c.Tbl = append(c.Tbl, tbl2.X())

	// place tbl1 inside tbl2/cell2
	tc2 := cell2.X()
	elts2 := wml.NewEG_BlockLevelElts()
	tc2.EG_BlockLevelElts = append(tc2.EG_BlockLevelElts, elts)
	c2 := wml.NewEG_ContentBlockContent()
	elts2.EG_ContentBlockContent = append(elts2.EG_ContentBlockContent, c2)
	c2.Tbl = append(c.Tbl, tbl1.X())

	doc.SaveToFile("recur.docx")
}
```

Reply:
Thanks.  It seems like a more serious issue when processing "arbitrary" input files (potentially from a malicious party) rather than when generating new files.

Is it possible to create a .docx, load and save back out which would cause this kind of problem?
Otherwise, I agree it should be treated as a separate issue.

Reply:
I don't think it is possible, unless there is a bug in the parser.

The input .docx file has to represent the actual layout of the data, so all nested tables would be nested in the xml text input. You could have an input .docx file with arbitrarily many nested tables, but for there to be an infinite number would require a file of infinite size.

I haven't thought a ton about the function to check this, but presumably there would only be a few potential places for circular references, including the one I'm making use of.

Reply:
OK.  One possible way to guard is also setting a maximum recursion depth.  For example in the bookmarks function, could pass an additional parameter depth int, which is incremented when calling itself recursively bookmarks(...,depth+1).

I will create another issue for the example you posted above.

We may also try running go-fuzz on this later.  That usually helps finding those kinds of errors too.

Reply:
Recursion depth is a good option. Whatever approach you decide to use on the save methods, I'm happy to use here as well.